### PR TITLE
Fix db.client.connection.count metric drifting over time

### PR DESCRIPTION
**Pull request by verdie-g**
## Problem

The `db.client.connection.count` UpDownCounter drifts in long-running applications, producing nonsensical values like large positive idle / negative used connection counts that never recover.

## Root cause

The dominant cause is a one-sided recording in the **async** `BlockingConnectionPool`, plus three smaller `±1` accounting bugs in the sync pool.

### Bug 0 — async `BlockingConnectionPool` never records the acquire (the actual drift)

`redis/asyncio/connection.py` — `BlockingConnectionPool.get_connection()` overrides the base `ConnectionPool.get_connection()` but only records `create_time` / `wait_time`; it **never records the acquire-side `connection.count` transition**. Meanwhile `BlockingConnectionPool.release()` (via `super().release()`) still records `USED -1 / IDLE +1`.

So every acquire/release cycle nets `USED -1, IDLE +1`. Under sustained traffic this drifts linearly and never recovers — matching the reported "+N idle / -N used" shape. It affects any app on the async `BlockingConnectionPool`, including Sentinel deployments through `SentinelConnectionPool` + `BlockingConnectionPool` (e.g. a `BlockingSentinelConnectionPool(SentinelConnectionPool, BlockingConnectionPool)` whose `get_connection` resolves to the buggy override). No fork is required.

Reproduced with an in-memory OTel `MeterProvider`: 1000 balanced acquire/release cycles on the async `BlockingConnectionPool` yield `used=-1000, idle=+1000`, while the base async `ConnectionPool` correctly nets to zero.

### Bug 1 — Phantom IDLE transitions (sync)
`ConnectionPool.make_connection()` and `BlockingConnectionPool.make_connection()` recorded `IDLE +1` for new connections, then `get_connection()` unconditionally recorded `IDLE -1`. Balanced in isolation, but creates unnecessary IDLE transitions that drift if any step is disrupted. The async pool already avoids this with an `is_created` check.

### Bug 2 — Wrong pool name on unowned connections (sync)
`release()` recorded `USED -1` against the literal `"unknown_pool"` when `owns_connection()` returned `False` (e.g. after a fork). The real pool's USED counter was never decremented.

### Bug 3 — Full queue silent drop (sync)
`BlockingConnectionPool.release()` swallowed both the `USED -1` and `IDLE +1` recordings when `put_nowait()` raised `Full`. The USED counter leaked +1 permanently and the dropped connection was never disconnected.

## Fix

| Bug | Fix |
|-----|-----|
| Async BlockingConnectionPool (Bug 0) | Record the acquire transition in `get_connection()` — new connections record `USED +1`, reused connections record `IDLE -1, USED +1` — mirroring the base `ConnectionPool` so it balances `release()`. |
| Phantom IDLE | Remove `IDLE +1` from `make_connection()`, add `is_created` check in `get_connection()`. Matches the async pool. |
| Wrong pool name | Use `get_pool_name(self)` instead of `"unknown_pool"`. |
| Full queue drop | In `except Full`, disconnect the connection, remove it from `_connections`, and record `USED -1`. |

## Tests

- `tests/test_asyncio/test_connection.py::TestAsyncBlockingConnectionPoolMetricCount` — new/reused/full-lifecycle (net-zero) recording for the async `BlockingConnectionPool`. These fail before Bug 0's fix (idle drift) and pass after.
- Sync metric-count tests in `tests/test_connection.py` covering the three sync bugs across `ConnectionPool` and `BlockingConnectionPool`.

---

> [!NOTE]
> **Medium Risk**
> Changes observability accounting and release edge cases (full queue, unowned connections) on core connection pools; behavior is test-backed but affects metrics operators rely on for capacity signals.
> 
> **Overview**
> Fixes **drift** in the `db.client.connection.count` UpDownCounter so IDLE/USED stay balanced under real pool traffic.
> 
> **Async `BlockingConnectionPool`** now records acquire-side metrics in `get_connection()` (new connections: `USED +1`; reused: `IDLE -1` / `USED +1`), matching `release()` and fixing linear drift (e.g. large negative USED / positive IDLE) seen on long-running async + Sentinel workloads.
> 
> **Sync pools** stop counting phantom IDLE on brand-new connections by removing `IDLE +1` from `make_connection()` and branching on `is_created` in `get_connection()`. **Unowned** connections (e.g. after fork) no longer emit `USED -1` on `release()`—fork cleanup stays in `reset()`/`__del__`. When **`BlockingConnectionPool.release()`** hits a full queue, the connection is disconnected, removed from `_connections`, and `USED -1` is recorded so metrics and later `reset()` do not double-decrement.
> 
> Adds **metric-count tests** for sync `ConnectionPool` / `BlockingConnectionPool` and async `BlockingConnectionPool`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 625115b645c6f5bc6c499c03dc8e565371d40deb. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by petyaslavova** [comment]
Hi @verdie-g, thank you for your contribution! Please ping me when the PR is ready for review :)

**Comment by verdie-g** [comment]
Hi @petyaslavova, it should be ready to review :)

**Comment by verdie-g** [comment]
@petyaslavova LGTY?

**Comment by petyaslavova** [comment]
> @petyaslavova LGTY?

Hey @verdie-g, I still haven't had a chance to review it :( Hopefully, in the following weeks, I will have some time to go through this one...

File: redis/connection.py [hunk 1]
```diff
@@ -3263,8 +3264,9 @@ def release(self, connection: "Connection") -> None:
                 # because the connection was not created by the pool.
                 # Still need to decrement USED since it was counted in get_connection()
                 connection.disconnect()
+                pool_name = get_pool_name(self)
```
Comment (petyaslavova) [inline redis/connection.py]: Following the forked-process case (the only place `owns_connection()` actually returns `False`), this `else` branch has two problems:

1. **Double decrement if it ever ran.** `self._checkpid()` at the top of `release()` triggers `reset()`, which clears `_in_use_connections` / `_available_connections` and already records `USED -in_use_count` — including *this* connection. So if the `else` executed afterward, its `USED -1` would decrement the same connection a second time.

2. **But it's unreachable on the fork path it's meant to handle** — so the fix here is on effectively dead code:
   - `ConnectionPool`: after `reset()` empties `_in_use_connections`, `self._in_use_connections.remove(connection)` raises `KeyError` and `release()` returns early, before `owns_connection()` is ever checked. (The `"unknown_pool" → get_pool_name` change is only reached by the new test, which sets `conn.pid = -1` without triggering `reset()` — a single-process state a real fork never produces.)
   - `BlockingConnectionPool`: it does enter the `else`, but `self.pool.put_nowait(None)` runs first, and `reset()` just refilled the queue with `None`s to capacity, so it raises `Full` (unhandled in this branch) before the `USED -1` line — and `release()` raises out.

   A side effect of the `ConnectionPool` early return: the inherited connection is neither returned to the pool nor `disconnect()`ed — it's silently abandoned.

Net: `reset()` already owns the fork-time accounting, so this branch is redundant with it and unreachable on the real fork path. This reinforces that the actual drift (and the abandoned-connection cleanup) needs to be handled in the fork path / `reset()` / `__del__`, not here.

**Review by petyaslavova — CHANGES_REQUESTED** [review]
Thanks for this — the three bugs are real and really need attention and fixing. 
But I don't think this PR resolves the problem in the description ("drift in long-running apps → +5M idle / -5M used"). 
All three fixes are `±1`-per-operation corrections; the symptom is a runaway, ever-growing drift, and that comes from a different mechanism the PR doesn't touch: **fork amplification.**

**Where the counter value actually lives.** The `db.client.connection.count` total is not held on the pool. It lives in the process-global `_metrics_collector` (`redis/observability/recorder.py`) and, behind it, in the OpenTelemetry SDK's aggregation storage. The pool only emits `±1` deltas tagged with `pool_name`, where `pool_name` is derived from `_pool_id`, which is generated once in `__init__` (`connection.py:2961`) and **never regenerated on fork** and the Redis server address.

**The amplification.** `ConnectionPool.reset()`'s metric-recording block only ever runs on fork — `__init__`'s call is skipped by the `hasattr(...)` guard, and the only other caller is `_checkpid()`. On fork it records `-len(connections)` for the *inherited* bookkeeping. `__del__` does the same, unguarded. So with a 100-connection pool forked into 10 workers, we emit ~`-1000` of decrements for 100 connections that were only ever created once, all under the parent's inherited `pool_name`. Over a long-running pre-fork server (gunicorn/uwsgi/celery) that recycles workers, this accumulates into the multi-million drift the PR is named after.

**Why it's wrong regardless of OTel setup.** Metric deltas are per-process. A connection's `+1` was recorded by the parent's meter; only the parent may record its `-1`. A forked child inherits the pool's bookkeeping but not the increments:
- If OTel is re-initialized in the child (the required pattern for pre-fork servers), the child's meter starts at a **zero baseline**, so the inherited pool's `reset()`/`__del__` `-100` is pure drift.
- If OTel is *not* re-initialized, the child inherits the value but OTel's exporter runs on a background thread that does not survive `fork()`, so the child never exports — telemetry is simply lost. (This is also the configuration the OTel spec calls unsupported.)

Either way, a forked process must not emit decrements for connections its own live meter never counted.

**Suggested direction (stays within the existing UpDownCounter — no metric-type change):**
1. `reset()` (fork path): drop the decrement recording entirely and regenerate `_pool_id`, so each worker reports an isolated series that nets to zero over its own lifecycle.
2. `__del__` (both pools): guard with `if getattr(self, "pid", None) != os.getpid(): return`, so a forked child that inherited connections but never counted them doesn't emit phantom decrements.
3. Keep the three fixes here — they're correct — but they're secondary to this.

**Ask:** could you confirm what actually reproduces the reported drift? If it's a forking deployment, this PR won't stop it, and I'd rather land the fork fix (which is the named problem) with these three folded in, plus a test that forks a real subprocess and asserts the aggregate counter returns to zero after teardown. If the drift was observed in a single-process app, please share that repro — that would point at a different root cause than any of the three addressed here.

File: redis/connection.py [hunk 2]
```diff
@@ -3653,8 +3656,9 @@ def release(self, connection):
                 connection.disconnect()
                 self.pool.put_nowait(None)
                 # Still need to decrement USED since it was counted in get_connection()
+                pool_name = get_pool_name(self)
                 record_connection_count(
-                    pool_name="unknown_pool",
+                    pool_name=pool_name,
```
[the remark this answers is not in the document]
Comment (verdie-g) [reply redis/connection.py]: Good catch on the region, though the exact mechanism is a bit different — the second `USED -1` never actually runs on the real fork path, so there's no double-decrement:

- `ConnectionPool.release()`: after a fork, `_checkpid()` → `reset()` empties `_in_use_connections`, so `self._in_use_connections.remove(connection)` raises `KeyError` and `release()` returns early — the `else` branch is never reached.
- `BlockingConnectionPool.release()`: `reset()` refills the queue to capacity with `None`, so the unowned branch's `self.pool.put_nowait(None)` raises `Full` **before** the `USED -1` line. `release()` raises out; the recording is skipped.

I verified this by simulating a fork (pool created in another pid, connection carrying the parent pid): `release()` raises `Full` and the net `USED` delta during `release()` is `-1` — just the one `reset()` records — not `-2`.

That said, the underlying point (and @petyaslavova's earlier note on the same lines) stands: this recording is redundant with `reset()`/`__del__`, which already own fork-time accounting, and it sits on unreachable/broken code. I've removed the `USED -1` from both unowned-release branches so fork-time accounting lives solely in `reset()`/`__del__` and can't double-count. Fixed in 7e1badcf.

**Comment by verdie-g** [comment]
Indeed my PR was not fixing the real bug. I traced it to a single-process async service using the async BlockingConnectionPool (Sentinel-backed), with no fork. I reproduced deterministically with an in-memory OTel MeterProvider: 1000 balanced cycles on the async BlockingConnectionPool give used=-1000, idle=+1000, while the base async ConnectionPool nets to zero. I've updated the PR.

**Review by petyaslavova — CHANGES_REQUESTED** [review]
Thanks for tracing this to the async `BlockingConnectionPool` — that acquire-side gap is a real deterministic drift, and the fix (plus the sync ±1 corrections and the full-queue leak) looks correct and restores sync/async parity. Two things before merge.

**One required change.** Deleting the `USED -1` from the unowned branch of `ConnectionPool.release()` (and the sync `BlockingConnectionPool` equivalent) leaks the `USED` counter on a non-fork path. When `owns_connection()` returns `False` for a same-process reason — the `SentinelConnectionPool` case after a master failover — `_checkpid()` no-ops, the connection was a real checkout (`USED +1` recorded under this pool), and it is then dropped without ever landing in `reset()`/`__del__`. So the `+1` is orphaned. The "redundant with `reset()`/`__del__`" reasoning only holds on the fork path (where `release()` returns early via `KeyError` and never reaches this branch). Could you record `USED -1` under `get_pool_name(self)` inside the `if connection.pid == self.pid:` block (which is safe on fork — pid mismatch skips it) and add a small Sentinel-failover regression test? That is essentially your original "Bug 2" `get_pool_name` fix, just guarded.

**One confirmation.** Could you confirm the reported "+5M idle / -5M used" production drift was this single-process async path (not a pre-fork server)? If so, I'll merge with the scope narrowed to what's fixed here and open a separate follow-up for fork amplification (`reset()`/`__del__` still emit decrements under the parent's `pool_name` since `_pool_id` isn't regenerated on fork, and `__del__` has no pid guard). Thanks again!
